### DISCUSSION TITLE
Openssl v3.0.8 + wpa_supplicant2.10 can not connect 802.1x network

### DISCUSSION THREAD

**Discussion by VimuVR3** [openssl/openssl#23687] [Q&A] [answered]
wpa_supplicant v2.10 with OpenSSL v3.0.8

**Issue description:**
I used supplicant2.10+openssl3.0.8 to test connecting to an 802.1x network, I got the following results:
PEAP+MSCHAPV2:Failed
PEAP+GTC:Passed

I want to use PEAP+MSCHAPV2 authentication with openssl3.0.8 for windows NPS servers where GTC is not supported (only MSCHAPV2 is supported by default).
I used supplicant2.10+openssl1.1.1t where it can connect to 802.1x using PEAP+MSCHAPV2

**Steps to reproduce:**
Connect to the 802.1x network and select the encryption mode PEAP + MSCHAPV2 with OpenSSL v3.0.8

**Observed behavior:**
connect timeout

**Expected behavior:**
connect ok

**Log snippet:**

> 16:57:53:437 wpa_supplicant: OpenSSL: RX ver=0x303 content_type=256 (TLS header info/)
> 16:57:53:445 wpa_supplicant: EAP-PEAP: received Phase 2: code=1 identifier=185 length=43
> 16:57:53:445 wpa_supplicant: EAP-PEAP: Phase 2 Request: type=26
> 16:57:53:456 wpa_supplicant: EAP-PEAP: Selected Phase 2 EAP vendor 0 method 26
> 16:57:53:470 wpa_supplicant: EAP-MSCHAPV2: RX identifier 185 mschapv2_id 185
> 16:57:53:471 wpa_supplicant: EAP-MSCHAPV2: Received challenge
> 16:57:53:481 wpa_supplicant: EAP-MSCHAPV2: Generating Challenge Response
> 16:57:53:481 wpa_supplicant: OpenSSL: EVP_DigestInit_ex failed: error:0308010C:digital envelope routines::unsupported

**Log snippet using eapol_test utility to test 802.1x authentication provided in wpa_supplicant:**

> EAP-MSCHAPV2: Generating Challenge Response
> Get randomness: len=16 entropy=0
> random from os_get_random - hexdump(len=16): 77 b5 40 38 12 e0 da 75 3c 96 41 67 9a 40 6a f5
> random_mix_pool - hexdump(len=20): 0d b9 b1 bf 70 7c bd fa 8b 8c 0a 46 d8 96 87 a4 8e 89 0d 7d
> random from internal pool - hexdump(len=16): 52 c7 66 0a bf 85 ed d3 d8 c1 5b 8c 5d 36 f0 8e
> mixed random - hexdump(len=16): 25 72 26 32 ad 65 37 a6 e4 57 1a eb c7 76 9a 7b
> MSCHAPV2: Identity - hexdump_ascii(len=5):
>      61 64 6d 69 6e                                    admin           
> MSCHAPV2: Username - hexdump_ascii(len=5):
>      61 64 6d 69 6e                                    admin           
> MSCHAPV2: auth_challenge - hexdump(len=16): 3e 04 b8 c6 6b 23 3d 40 cb bf 55 7b e4 b2 85 d9
> MSCHAPV2: peer_challenge - hexdump(len=16): 25 72 26 32 ad 65 37 a6 e4 57 1a eb c7 76 9a 7b
> MSCHAPV2: username - hexdump_ascii(len=5):
>      61 64 6d 69 6e                                    admin           
> MSCHAPV2: password - hexdump_ascii(len=8):
>      70 61 73 73 77 6f 72 64                           password        
> OpenSSL: EVP_DigestInit_ex failed: error:0308010C:digital envelope routines::unsupported
> EAP-MSCHAPV2: Failed to derive response
> EAP: method process -> ignore=FALSE methodState=MAY_CONT decision=FAIL eapRespData=0
> EAP: EAP entering state SEND_RESPONSE
> EAP: No eapRespData available
> EAP: EAP entering state IDLE
> EAPOL test timed out
> EAPOL: EAP key not available
> EAPOL: EAP Session-Id not available
> WPA: Clear old PMK and PTK
> EAP: deinitialize previously used EAP method (25, PEAP) at EAP deinit
> MPPE keys OK: 0  mismatch: 1
> FAILURE

Further, when debugging it is found NULL value is returned in the following code snippet of **crypto/evp/digest.c**
In function **evp_md_init_internal**, **EVP_MD_fetch** returns a **NULL** value below which it is returned inside NULL check.
```
        /* The NULL digest is a special case */
        EVP_MD *provmd = EVP_MD_fetch(NULL,
                                      type->type != NID_undef ? OBJ_nid2sn(type->type)
                                                              : "NULL", "");

        if (provmd == NULL) {
            ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
            return 0;
        }
```

**Comment by beldmit** [comment]
Probably you need to load the legacy provider as the digest used for authorization is insecure and is absent in the default provider. See https://bugzilla.redhat.com/show_bug.cgi?id=2072070 for details

**Reply by VimuVR3** [reply]
Anything related to eapol_test specifically? as I'm using eapol_test provided with wpa_supplicant for testing purposes in my network device (Access Point).
I tried making the patch changes in hostapd provided here https://bugzilla.redhat.com/show_bug.cgi?id=2072070, which didn't help me out with eapol_test.

**Comment by VimuVR3** [comment]
Further, came to know that in https://bugzilla.redhat.com/show_bug.cgi?id=2072070, the SSL error code is different to the issue reported here.

Reference link log:
    Feb 23 10:58:47 dmk-laptop-21 wpa_supplicant[1404]: OpenSSL: openssl_handshake - SSL_connect error:0A000152:SSL routines::unsafe legacy renegotiation disabled
    Feb 23 10:58:48 dmk-laptop-21 wpa_supplicant[1404]: wlp147s0: CTRL-EVENT-EAP-FAILURE EAP authentication failed

Issue reported:
16:57:53:481 wpa_supplicant: OpenSSL: EVP_DigestInit_ex failed: error:0308010C:digital envelope routines::unsupported

or

OpenSSL: EVP_DigestInit_ex failed: error:0308010C:digital envelope routines::unsupported
EAP-MSCHAPV2: Failed to derive response

**Reply by beldmit** [reply]
Could you please find out which digest is not found by OpenSSL?

**Reply by VimuVR3** [reply]
I didn't use any digest specifically. might it be selecting a default one in noop case?

Freeradius server logs:
(37)     [preprocess] = ok
(37)     [chap] = noop
(37)     [mschap] = noop
(37)     [digest] = noop

**Comment by beldmit** [comment]
Again, could you please try enabling the legacy provider? I think the digest or cipher (or both) are not included into the default provider as MSCHAP2 was implemented in 1998 or so, according to Wikipedia

**Reply by VimuVR3** [reply]
Could you please share the steps to enable the leagcy provider?

Let me follow the same and try to figure it out.

**Reply by beldmit** [reply, marked as answer]
Try editing the corresponding part of OpenSSL config (usually somewhere in /etc) to make it look like

```
[provider_sect]
default = default_sect
legacy = legacy_sect

[default_sect]
activate = 1

[legacy_sect]
activate = 1
```

**Reply by VimuVR3** [reply]
Found the conf file here, /etc/ssl/openssl.cnf 

Would it work if the legacy config changes are done in runtime? Still the issue persists the same.

I haven't done a compilation dependency check along with openssl.cnf changes yet.

**Reply by beldmit** [reply]
If I correctly understand your question, you don't have to recompile openssl to enable the legacy provider

**Reply by VimuVR3** [reply]
In that case, I added the mentioned changes to the conf file and verified the auth test.
Still the same issue is observed.

**Reply by beldmit** [reply]
Could you please point at the code calling the `EVP_DigestInit_ex`?

**Reply by VimuVR3** [reply]
I could see that `EVP_DigestInit_ex` is called from `EVP_Digest` at the time of issue

In crypto/evp/digest.c

```
int EVP_Digest(const void *data, size_t count,
               unsigned char *md, unsigned int *size, const EVP_MD *type,
               ENGINE *impl)
{
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
    int ret; 

    if (ctx == NULL)
        return 0;
    EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_ONESHOT);
    ret = EVP_DigestInit_ex(ctx, type, impl)
        && EVP_DigestUpdate(ctx, data, count)
        && EVP_DigestFinal_ex(ctx, md, size);
    EVP_MD_CTX_free(ctx);

    return ret; 
}
```

**Reply by beldmit** [reply]
Unfortunately that's not the code I'm looking for. What digest is the application looking for?

**Reply by VimuVR3** [reply]
From the server, these logs are observed for selected Cipher-suite which is sent in a Access-Challenge packet.

> (7) TLS-Session-Cipher-Suite = "ECDHE-RSA-CHACHA20-POLY1305"
> (7) TLS-Session-Version = "TLS 1.2"

Possibly, sha256 could be the digest? not sure.

**Reply by VimuVR3** [reply]
Moreover, found that `EVP_Digest` is getting called from `ossl_asn1_item_digest_ex` in crypto/asn1/a_digest.c

```
int ossl_asn1_item_digest_ex(const ASN1_ITEM *it, const EVP_MD *md, void *asn,
                             unsigned char *data, unsigned int *len,
                             OSSL_LIB_CTX *libctx, const char *propq)
{
    int i, ret = 0;
    unsigned char *str = NULL;
    EVP_MD *fetched_md = (EVP_MD *)md;

    i = ASN1_item_i2d(asn, &str, it);
    if (i < 0 || str == NULL)
        return 0;

    if (EVP_MD_get0_provider(md) == NULL) {
#if !defined(OPENSSL_NO_ENGINE)
        ENGINE *tmpeng = ENGINE_get_digest_engine(EVP_MD_get_type(md));

        if (tmpeng != NULL)
            ENGINE_finish(tmpeng);
        else
#endif
            fetched_md = EVP_MD_fetch(libctx, EVP_MD_get0_name(md), propq);
    }   
    if (fetched_md == NULL)
        goto err;

    ret = EVP_Digest(str, i, data, len, fetched_md, NULL);
err:
    OPENSSL_free(str);
    if (fetched_md != md) 
        EVP_MD_free(fetched_md);
    return ret;
}
```

**Reply by beldmit** [reply]
I think that either the pieces of code you are looking for are outside of OpenSSL (in the calling application) or using some certificate provided by application.

**Reply by VimuVR3** [reply]
Just in case, the same certificate was used by OpenSSL v1.1.1t (earlier version), where it worked fine.

**Reply by beldmit** [reply]
Yes. OpenSSL 3 has stricter defaults (e.g. SHA1 in signatures may not work by default)

**Reply by VimuVR3** [reply]
Is it possible to include unsupported ones to the defaults as per requirement?

**Reply by beldmit** [reply]
In most cases it is doable via configuration. But currently we don't know which exactly digest can't be found.

**Reply by VimuVR3** [reply]
> From the server, these logs are observed for selected Cipher-suite which is sent in a Access-Challenge packet.
> 
> (7) TLS-Session-Cipher-Suite = "ECDHE-RSA-CHACHA20-POLY1305"
> (7) TLS-Session-Version = "TLS 1.2"
> 
> Possibly, sha256 could be the digest? not sure.

Did you come across this earlier comment? SHA256 might be the digest used for ECDHE-RSA-CHACHA20-POLY1305. If it might be, could you share conf file changes to include support for this digest?

**Reply by beldmit** [reply]
No. SHA256 can't cause this problem. I can't imagine a configuration when SHA256 is not available and any other openssl application works.

**Reply by VimuVR3** [reply]
Is it possible to get and print the digest used by the application in OpenSSL codes?
`fetched_md = EVP_MD_fetch(libctx, EVP_MD_get0_name(md), propq);`
EVP_MD_get0_name(md) if it requires to print the md arg here in ossl_asn1_item_digest_ex or any of the digest handling codes.

**Reply by beldmit** [reply]
Yes, you can try add printing the information there.

**Reply by VimuVR3** [reply]
I managed to trace the path at the issue state partially.

`hmac_setkey` --> `HMAC_Init_ex` --> `EVP_DigestInit_ex`

where in `hmac_setkey` function (providers/implementations/macs/hmac_prov.c), digest is taken from `ossl_prov_digest_md` as in the following code snippet.

`digest = ossl_prov_digest_md(&macctx->digest);`

**Reply by VimuVR3** [reply]
In providers/implementations/macs/hmac_prov.c, at the function `hmac_setkey`
digest = ossl_prov_digest_md(&macctx->digest);

EVP_MD *digest is passed to `EVP_DigestInit_ex`

I printed the digest type in `hmac_setkey` by using EVP_MD_name(digest) where it is **SHA2-256**

EVP_MD_name(digest) = SHA2-256

In crypto/evp/digest.c, at the function `evp_md_init_internal` where EVP_MD_fetch is null and returned an error, I printed the EVP_MD *type in `evp_md_init_internal` (which is EVP_MD *digest passed as an arg from `hmac_setkey`) where it is referred as **MD4**

EVP_MD_name(type) = MD4

The digest type is changed from **SHA2-256** to **MD4** between `hmac_setkey` --> `HMAC_Init_ex` --> `EVP_DigestInit_ex` --> `evp_md_init_internal`

Is MD4 digest not supported in the defaults? or is this conversion causing the issue?

**Reply by beldmit** [reply]
MD4 is not supported in the default provider. It is supported in the legacy provider.

**Reply by VimuVR3** [reply]
But legacy provider changes in openssl.cnf is not resolving the state.

**Reply by beldmit** [reply]
Could you please verify that your application uses the system config and loads the legacy provider? Under linux it can be done via `strace` app

**Reply by VimuVR3** [reply]
Could you please provide a way to include legacy provider changes to the openssl.cnf file while compiling libopenssl?

**Reply by beldmit** [reply]
I don't think there is a compile option for it.

**Reply by VimuVR3** [reply]
I meant to include the legacy provider changes while generating the openssl.cnf by default.

**Reply by beldmit** [reply]
I understand but unfortunately it doesn't change my answer

**Reply by VimuVR3** [reply]
I tried to include the legacy provider changes with a patch file for openssl.cnf and recompiled openssl which reflected the changes very well.
Similarly, compiled the eapol_test application of wpa_supplicant where I expected it to work fine without the issue with the recompiled libopenssl legacy provider.

But still the issue persisted. Is there any way to include support for MD4 digest at the code level?

**Reply by VimuVR3** [reply]
I'm not able to activate the legacy provider with openssl.cnf

root@Vimu:~# openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.0.8
    status: active
    
 Even after the change, only default provider is active

**Reply by beldmit** [reply]
Could you please provide your config file?

**Reply by VimuVR3** [reply]
Please find the attached config file for reference.

**Reply by beldmit** [reply]
Don't see, sorry

**Reply by VimuVR3** [reply]
My bad, it is not getting uploaded here. Let me write it over the comment.

```
#
# OpenSSL example configuration file.
# See doc/man5/config.pod for more info.
#
# This is mostly being used for generation of certificate requests,
# but may be used for auto loading of providers

# Note that you can include other files from the main configuration
# file using the .include directive.
#.include filename

# This definition stops the following lines choking if HOME isn't
# defined.
HOME			= .

 # Use this in order to automatically load providers.
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

# Extra OBJECT IDENTIFIER info:
# oid_file       = $ENV::HOME/.oid
oid_section = new_oids

# To use this configuration file with the "-extfile" option of the
# "openssl x509" utility, name here the section containing the
# X.509v3 extensions to use:
# extensions		=
# (Alternatively, use a configuration file that has only
# X.509v3 extensions in its main [= default] section.)

openssl_conf=openssl_conf

[openssl_conf]
engines=engines

[engines]
# To enable an engine, install the package, and uncomment it here:
#devcrypto=devcrypto
#afalg=afalg
#padlock=padlock

[afalg]
# Leave this alone and configure algorithms with CIPERS/DIGESTS below
default_algorithms = ALL

# The following commands are only available if using the alternative
# (sync) AFALG engine
# Configuration commands:
# Run 'openssl engine -t -c -vv -pre DUMP_INFO devcrypto' to see a
# list of supported algorithms, along with their driver, whether they
# are hw accelerated or not, and the engine's configuration commands.

# USE_SOFTDRIVERS: specifies whether to use software (not accelerated)
# drivers (0=use only accelerated drivers, 1=allow all drivers, 2=use
# if acceleration can't be determined) [default=2]
#USE_SOFTDRIVERS = 2

# CIPHERS: either ALL, NONE, NO_ECB (all except ECB-mode) or a
# comma-separated list of ciphers to enable [default=NO_ECB]
# Starting in 1.2.0, if you use a cipher list, each cipher may be
# followed by a colon (:) and the minimum request length to use
# AF_ALG drivers for that cipher; smaller requests are processed by
# softare; a negative value will use the default for that cipher
#CIPHERS=AES-128-CBC:1024, AES-256-CBC:768, DES-EDE3-CBC:0

# DIGESTS: either ALL, NONE, or a comma-separated list of digests to
# enable [default=NONE]
# It is strongly recommended not to enable digests; their performance
# is poor, and there are many cases in which they will not work,
# especially when calling fork with open crypto contexts.  Openssh,
# for example, does this, and you may not be able to login.
#DIGESTS = NONE

[devcrypto]
# Leave this alone and configure algorithms with CIPERS/DIGESTS below
default_algorithms = ALL

# Configuration commands:
# Run 'openssl engine -t -c -vv -pre DUMP_INFO devcrypto' to see a
# list of supported algorithms, along with their driver, whether they
# are hw accelerated or not, and the engine's configuration commands.

# USE_SOFTDRIVERS: specifies whether to use software (not accelerated)
# drivers (0=use only accelerated drivers, 1=allow all drivers, 2=use
# if acceleration can't be determined) [default=2]
#USE_SOFTDRIVERS = 2

# CIPHERS: either ALL, NONE, or a comma-separated list of ciphers to
# enable [default=ALL]
# It is recommended to disable the ECB ciphers; in most cases, it will
# only be used for PRNG, in small blocks, where performance is poor,
# and there may be problems with apps forking with open crypto
# contexts, leading to failures.  The CBC ciphers work well:
#CIPHERS=DES-CBC, DES-EDE3-CBC, AES-128-CBC, AES-192-CBC, AES-256-CBC

# DIGESTS: either ALL, NONE, or a comma-separated list of digests to
# enable [default=NONE]
# It is strongly recommended not to enable digests; their performance
# is poor, and there are many cases in which they will not work,
# especially when calling fork with open crypto contexts.  Openssh,
# for example, does this, and you may not be able to login.
#DIGESTS = NONE

[padlock]
default_algorithms = ALL

[ new_oids ]
# We can add new OIDs in here for use by 'ca', 'req' and 'ts'.
# Add a simple OID like this:
# testoid1=1.2.3.4
# Or use config file substitution like this:
# testoid2=${testoid1}.5.6

# Policies used by the TSA examples.
tsa_policy1 = 1.2.3.4.1
tsa_policy2 = 1.2.3.4.5.6
tsa_policy3 = 1.2.3.4.5.7

# For FIPS
# Optionally include a file that is generated by the OpenSSL fipsinstall
# application. This file contains configuration data required by the OpenSSL
# fips provider. It contains a named section e.g. [fips_sect] which is
# referenced from the [provider_sect] below.
# Refer to the OpenSSL security policy for more information.
# .include fipsmodule.cnf

[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
default = default_sect
legacy = legacy_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
# fips = fips_sect

# If no providers are activated explicitly, the default one is activated implicitly.
# See man 7 OSSL_PROVIDER-default for more details.
#
# If you add a section explicitly activating any other provider(s), you most
# probably need to explicitly activate the default provider, otherwise it
# becomes unavailable in openssl.  As a consequence applications depending on
# OpenSSL may not work correctly which could lead to significant system
# problems including inability to remotely access the system.
[default_sect]
activate = 1

[legacy_sect]
activate = 1


####################################################################
[ ca ]
default_ca	= CA_default		# The default ca section

####################################################################
[ CA_default ]

dir		= ./demoCA		# Where everything is kept
certs		= $dir/certs		# Where the issued certs are kept
crl_dir		= $dir/crl		# Where the issued crl are kept
database	= $dir/index.txt	# database index file.
#unique_subject	= no			# Set to 'no' to allow creation of
					# several certs with same subject.
new_certs_dir	= $dir/newcerts		# default place for new certs.

certificate	= $dir/cacert.pem 	# The CA certificate
serial		= $dir/serial 		# The current serial number
crlnumber	= $dir/crlnumber	# the current crl number
					# must be commented out to leave a V1 CRL
crl		= $dir/crl.pem 		# The current CRL
private_key	= $dir/private/cakey.pem# The private key

x509_extensions	= usr_cert		# The extensions to add to the cert

# Comment out the following two lines for the "traditional"
# (and highly broken) format.
name_opt 	= ca_default		# Subject Name options
cert_opt 	= ca_default		# Certificate field options

# Extension copying option: use with caution.
# copy_extensions = copy

# Extensions to add to a CRL. Note: Netscape communicator chokes on V2 CRLs
# so this is commented out by default to leave a V1 CRL.
# crlnumber must also be commented out to leave a V1 CRL.
# crl_extensions	= crl_ext

default_days	= 365			# how long to certify for
default_crl_days= 30			# how long before next CRL
default_md	= default		# use public key default MD
preserve	= no			# keep passed DN ordering

# A few difference way of specifying how similar the request should look
# For type CA, the listed attributes must be the same, and the optional
# and supplied fields are just that :-)
policy		= policy_match

# For the CA policy
[ policy_match ]
countryName		= match
stateOrProvinceName	= match
organizationName	= match
organizationalUnitName	= optional
commonName		= supplied
emailAddress		= optional

# For the 'anything' policy
# At this point in time, you must list all acceptable 'object'
# types.
[ policy_anything ]
countryName		= optional
stateOrProvinceName	= optional
localityName		= optional
organizationName	= optional
organizationalUnitName	= optional
commonName		= supplied
emailAddress		= optional

####################################################################
[ req ]
default_bits		= 2048
default_keyfile 	= privkey.pem
distinguished_name	= req_distinguished_name
attributes		= req_attributes
x509_extensions	= v3_ca	# The extensions to add to the self signed cert

# Passwords for private keys if not present they will be prompted for
# input_password = secret
# output_password = secret

# This sets a mask for permitted string types. There are several options.
# default: PrintableString, T61String, BMPString.
# pkix	 : PrintableString, BMPString (PKIX recommendation before 2004)
# utf8only: only UTF8Strings (PKIX recommendation after 2004).
# nombstr : PrintableString, T61String (no BMPStrings or UTF8Strings).
# MASK:XXXX a literal mask value.
# WARNING: ancient versions of Netscape crash on BMPStrings or UTF8Strings.
string_mask = utf8only

# req_extensions = v3_req # The extensions to add to a certificate request

[ req_distinguished_name ]
countryName			= Country Name (2 letter code)
countryName_default		= AU
countryName_min			= 2
countryName_max			= 2

stateOrProvinceName		= State or Province Name (full name)
stateOrProvinceName_default	= Some-State

localityName			= Locality Name (eg, city)

0.organizationName		= Organization Name (eg, company)
0.organizationName_default	= Internet Widgits Pty Ltd

# we can do this but it is not needed normally :-)
#1.organizationName		= Second Organization Name (eg, company)
#1.organizationName_default	= World Wide Web Pty Ltd

organizationalUnitName		= Organizational Unit Name (eg, section)
#organizationalUnitName_default	=

commonName			= Common Name (e.g. server FQDN or YOUR name)
commonName_max			= 64

emailAddress			= Email Address
emailAddress_max		= 64

# SET-ex3			= SET extension number 3

[ req_attributes ]
challengePassword		= A challenge password
challengePassword_min		= 4
challengePassword_max		= 20

unstructuredName		= An optional company name

[ usr_cert ]

# These extensions are added when 'ca' signs a request.

# This goes against PKIX guidelines but some CAs do it and some software
# requires this to avoid interpreting an end user certificate as a CA.

basicConstraints=CA:FALSE

# This is typical in keyUsage for a client certificate.
# keyUsage = nonRepudiation, digitalSignature, keyEncipherment

# PKIX recommendations harmless if included in all certificates.
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid,issuer

# This stuff is for subjectAltName and issuerAltname.
# Import the email address.
# subjectAltName=email:copy
# An alternative to produce certificates that aren't
# deprecated according to PKIX.
# subjectAltName=email:move

# Copy subject details
# issuerAltName=issuer:copy

# This is required for TSA certificates.
# extendedKeyUsage = critical,timeStamping
# Extended Key Usage Extension.
[ ext_usage ]
extendedKeyUsage=serverAuth

[ v3_req ]

# Extensions to add to a certificate request

basicConstraints = CA:FALSE
keyUsage = nonRepudiation, digitalSignature, keyEncipherment

[ v3_ca ]


# Extensions for a typical CA


# PKIX recommendation.

subjectKeyIdentifier=hash

authorityKeyIdentifier=keyid:always,issuer

basicConstraints = critical,CA:true

# Key usage: this is typical for a CA certificate. However since it will
# prevent it being used as an test self-signed certificate it is best
# left out by default.
# keyUsage = cRLSign, keyCertSign

# Include email address in subject alt name: another PKIX recommendation
# subjectAltName=email:copy
# Copy issuer details
# issuerAltName=issuer:copy

# DER hex encoding of an extension: beware experts only!
# obj=DER:02:03
# Where 'obj' is a standard or added object
# You can even override a supported extension:
# basicConstraints= critical, DER:30:03:01:01:FF

[ crl_ext ]

# CRL extensions.
# Only issuerAltName and authorityKeyIdentifier make any sense in a CRL.

# issuerAltName=issuer:copy
authorityKeyIdentifier=keyid:always

[ proxy_cert_ext ]
# These extensions should be added when creating a proxy certificate

# This goes against PKIX guidelines but some CAs do it and some software
# requires this to avoid interpreting an end user certificate as a CA.

basicConstraints=CA:FALSE

# This is typical in keyUsage for a client certificate.
# keyUsage = nonRepudiation, digitalSignature, keyEncipherment

# PKIX recommendations harmless if included in all certificates.
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid,issuer

# This stuff is for subjectAltName and issuerAltname.
# Import the email address.
# subjectAltName=email:copy
# An alternative to produce certificates that aren't
# deprecated according to PKIX.
# subjectAltName=email:move

# Copy subject details
# issuerAltName=issuer:copy

# This really needs to be in place for it to be a proxy certificate.
proxyCertInfo=critical,language:id-ppl-anyLanguage,pathlen:3,policy:foo

####################################################################
[ tsa ]

default_tsa = tsa_config1	# the default TSA section

[ tsa_config1 ]

# These are used by the TSA reply generation only.
dir		= ./demoCA		# TSA root directory
serial		= $dir/tsaserial	# The current serial number (mandatory)
crypto_device	= builtin		# OpenSSL engine to use for signing
signer_cert	= $dir/tsacert.pem 	# The TSA signing certificate
					# (optional)
certs		= $dir/cacert.pem	# Certificate chain to include in reply
					# (optional)
signer_key	= $dir/private/tsakey.pem # The TSA private key (optional)
signer_digest  = sha256			# Signing digest to use. (Optional)
default_policy	= tsa_policy1		# Policy if request did not specify it
					# (optional)
other_policies	= tsa_policy2, tsa_policy3	# acceptable policies (optional)
digests     = sha1, sha256, sha384, sha512  # Acceptable message digests (mandatory)
accuracy	= secs:1, millisecs:500, microsecs:100	# (optional)
clock_precision_digits  = 0	# number of digits after dot. (optional)
ordering		= yes	# Is ordering defined for timestamps?
				# (optional, default: no)
tsa_name		= yes	# Must the TSA name be included in the reply?
				# (optional, default: no)
ess_cert_id_chain	= no	# Must the ESS cert id chain be included?
				# (optional, default: no)
ess_cert_id_alg		= sha1	# algorithm to compute certificate
				# identifier (optional, default: sha1)

[insta] # CMP using Insta Demo CA
# Message transfer
server = pki.certificate.fi:8700
# proxy = # set this as far as needed, e.g., http://192.168.1.1:8080
# tls_use = 0
path = pkix/

# Server authentication
recipient = "/C=FI/O=Insta Demo/CN=Insta Demo CA" # or set srvcert or issuer
ignore_keyusage = 1 # potentially needed quirk
unprotected_errors = 1 # potentially needed quirk
extracertsout = insta.extracerts.pem

# Client authentication
ref = 3078 # user identification
secret = pass:insta # can be used for both client and server side

# Generic message options
cmd = ir # default operation, can be overridden on cmd line with, e.g., kur

# Certificate enrollment
subject = "/CN=openssl-cmp-test"
newkey = insta.priv.pem
out_trusted = insta.ca.crt
certout = insta.cert.pem

[pbm] # Password-based protection for Insta CA
# Server and client authentication
ref = $insta::ref # 3078
secret = $insta::secret # pass:insta

[signature] # Signature-based protection for Insta CA
# Server authentication
trusted = insta.ca.crt # does not include keyUsage digitalSignature

# Client authentication
secret = # disable PBM
key = $insta::newkey # insta.priv.pem
cert = $insta::certout # insta.cert.pem

[ir]
cmd = ir

[cr]
cmd = cr

[kur]
# Certificate update
cmd = kur
oldcert = $insta::certout # insta.cert.pem

[rr]
# Certificate revocation
cmd = rr
oldcert = $insta::certout # insta.cert.pem
```

**Reply by beldmit** [reply]
I think the lines `openssl_conf = openssl_init` and `openssl_conf = openssl_conf` are in a contradiction. So the 1st is ignored, and provider sections also are. And, when no providers are configured, openssl loads the default one automatically.

**Reply by VimuVR3** [reply]
Okay, let me try to include `providers = provider_sect` in `[openssl_conf]` as below.

```
[openssl_conf]
engines=engines
providers = provider_sect
```
Would it work then?

Note: I tried runtime which didn't work, will try to do with a compile.

**Reply by beldmit** [reply]
Yes, it should work (and you can check it via  `openssl list -providers`

**Reply by VimuVR3** [reply]
I think legacy.so is not loaded in my device

```
root@Vimu:~# echo -n "test" | openssl md4 -provider legacy
md4: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
B474F676:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/usr/lib/ossl-modules/legacy.so): Error loading shared library /usr/lib/ossl-modules/legacy.so: No such file or directory
B474F676:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
B474F676:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:912:name=legacy
```
```
root@Vimu:~# echo -n "test" | openssl md4
Error setting digest
B4D4FC76:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (MD4 : 50), Properties ()
B4D4FC76:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:267:

```
Could you please share the changes need to be done in Makefile to include legacy.so?

**Reply by beldmit** [reply]
`./config shared` does the trick.
Could you please run `openssl version -a` and see what is installed in your MODULESDIR?

**Reply by VimuVR3** [reply]
Please find the below details

```
root@Vimu:~# openssl version -a
OpenSSL 3.0.8 7 Feb 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)
built on: Wed Mar 13 10:02:04 2024 UTC
platform: linux-arm-openwrt
options:  bn(64,32)
compiler: arm-openwrt-linux-muslgnueabi-gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -Os -pipe -march=armv7-a -mtune=cortex-a7 -fno-caller-saves -Wa,--noexecstack -fhonour-copts -mfloat-abi=soft -fPIC -fstack-protector -D_FORTIFY_SOURCE=2 -Wl,-z,now -Wl,-z,relro -fPIC -ffunction-sections -fdata-sections -specs=/home/docker/workspace/smb-wifi/sdk/qc/kale/ipq60xx/qsdk/include/hardened-ld-pie.specs -znow -zrelro -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -DOPENSSL_PREFER_CHACHA_OVER_GCM -DOPENSSL_SMALL_FOOTPRINT
OPENSSLDIR: "/etc/ssl"
ENGINESDIR: "/usr/lib/engines-1.1"
MODULESDIR: "/usr/lib/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_armcap=0x3d

```

**Reply by beldmit** [reply]
`ls /usr/lib/ossl-modules` ?

**Reply by VimuVR3** [reply]
No dir named as /usr/lib/ossl-modules

libssl so files are in /usr/lib/
There is no legacy.so and dir /usr/lib/ossl-modules as well

root@Vimu:~# ls -l /usr/lib/libssl.so.*
-rw-r--r--    1 root     root        458648 Mar 13  2024 /usr/lib/libssl.so.1.1
-rw-r--r--    1 root     root        458648 Mar 13  2024 /usr/lib/libssl.so.3

**Reply by beldmit** [reply]
Then you don't have a legacy provider. If it is built and installed, it should be in the MODULESDIR.
Sorry, I don't know details of your system and can't help you much.

**Reply by beldmit** [reply]
As it's openwrt, probably they didn't build it at all.

**Reply by VimuVR3** [reply]
Okay, let me check with the Makefile to install and include legacy.so first.
Later, do the legacy provider activation and see if the issue is observed still.

**Reply by VimuVR3** [reply]
@beldmit 

After installing legacy.so it is able to activate legacy provider with openssl.cnf

```
root@Vimu:~# openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.0.8
    status: active
  legacy
    name: OpenSSL Legacy Provider
    version: 3.0.8
    status: active

```
**MD4** digest is working fine for **PEAP + MSCHAPV2** with OpenSSLv3.0.8 with legacy provider.

Thanks for your support and patience so far.
